File: ada_assign/code/javaLang/MergeSortStep.java
package javaLang;

import java.io.*;
import java.util.*;

public class MergeSortStep {
    public static void main(String[] args) throws IOException {
        if (args.length != 3) {
            System.out.println("Usage: java MergeSortStep <filename> <startRow> <endRow>");
            return;
        }

        String filename = args[0];
        int startRow = Integer.parseInt(args[1]);
        int endRow = Integer.parseInt(args[2]);

        List<String> values = new ArrayList<>();

        try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            String line;
            int lineNo = 0;
            while ((line = reader.readLine()) != null) {
                // Adjusted to treat startRow and endRow as 1-indexed
                if (lineNo + 1 >= startRow && lineNo + 1 <= endRow) {
                    values.add(line.trim());
                }
                lineNo++;
            }
        }

        List<String> steps = new ArrayList<>();
        // Add initial state
        steps.add(convertToSlashFormat(values));

        mergeSort(values, steps, 0, values.size() - 1);

        try (BufferedWriter writer = new BufferedWriter(
                new FileWriter("./output/merge_sort_step_" + startRow + "_" + endRow + ".txt"))) {
            for (String step : steps) {
                writer.write(step);
                writer.newLine();
            }
        }
    }

    static void mergeSort(List<String> arr, List<String> steps, int left, int right) {
        if (left >= right)
            return;

        int mid = left + (right - left) / 2;

        // Recursively sort both halves
        mergeSort(arr, steps, left, mid);
        mergeSort(arr, steps, mid + 1, right);

        // Merge the sorted halves
        merge(arr, steps, left, mid, right);
    }

    static void merge(List<String> arr, List<String> steps, int left, int mid, int right) {
        List<String> leftArr = new ArrayList<>();
        List<String> rightArr = new ArrayList<>();

        for (int i = left; i <= mid; i++) {
            leftArr.add(arr.get(i));
        }
        for (int i = mid + 1; i <= right; i++) {
            rightArr.add(arr.get(i));
        }

        int i = 0, j = 0, k = left;

        while (i < leftArr.size() && j < rightArr.size()) {
            if (compare(leftArr.get(i), rightArr.get(j)) <= 0) {
                arr.set(k, leftArr.get(i));
                i++;
            } else {
                arr.set(k, rightArr.get(j));
                j++;
            }
            k++;
        }

        while (i < leftArr.size()) {
            arr.set(k, leftArr.get(i));
            i++;
            k++;
        }

        while (j < rightArr.size()) {
            arr.set(k, rightArr.get(j));
            j++;
            k++;
        }

        // Save step after each merge
        steps.add(convertToSlashFormat(arr));
    }

    static int compare(String a, String b) {
        int aNum = Integer.parseInt(a.split(",")[0]);
        int bNum = Integer.parseInt(b.split(",")[0]);
        return Integer.compare(aNum, bNum);
    }

    static String convertToSlashFormat(List<String> list) {
        List<String> formatted = new ArrayList<>();
        for (String entry : list) {
            String[] parts = entry.split(",");
            formatted.add(parts[0] + "/" + parts[1]);
        }
        return "[" + String.join(", ", formatted) + "]";
    }
}

File: ada_assign/code/pythonLang/merge_sort_step.py
import sys
import os


def convert_to_slash_format(arr):
    """
    Convert array to comma-separated list where each element is in number/text (slash) format.
    Example: [123/foo, 456/bar]
    """
    converted = [item.replace(",", "/") for item in arr]
    return "[" + ", ".join(converted) + "]"


def compare(s1, s2):
    """
    Compare two strings by their numeric part.
    """
    val1 = int(s1.split(",")[0])
    val2 = int(s2.split(",")[0])
    return val1 - val2


def merge(arr, low, mid, high, steps):
    """
    Merge function with logging after each full merge.
    """
    left = arr[low : mid + 1]
    right = arr[mid + 1 : high + 1]

    i = j = 0
    k = low

    while i < len(left) and j < len(right):
        if compare(left[i], right[j]) <= 0:
            arr[k] = left[i]
            i += 1
        else:
            arr[k] = right[j]
            j += 1
        k += 1

    while i < len(left):
        arr[k] = left[i]
        i += 1
        k += 1

    while j < len(right):
        arr[k] = right[j]
        j += 1
        k += 1

    # Log current state of full array
    steps.append(convert_to_slash_format(arr))


def merge_sort(arr, low, high, steps):
    """
    Recursive MergeSort with step logging after each merge.
    """
    if low < high:
        mid = (low + high) // 2
        merge_sort(arr, low, mid, steps)
        merge_sort(arr, mid + 1, high, steps)
        merge(arr, low, mid, high, steps)


def main():
    if len(sys.argv) != 4:
        print("Usage: python merge_sort_step.py <filename> <startRow> <endRow>")
        return

    filename = sys.argv[1]
    start_row = int(sys.argv[2]) - 1
    end_row = int(sys.argv[3]) - 1

    data = []

    try:
        with open(filename, "r") as file:
            for i, line in enumerate(file):
                if start_row <= i <= end_row:
                    line = line.strip()
                    if line:
                        data.append(line)

        steps = []
        steps.append(convert_to_slash_format(data))  # initial state

        merge_sort(data, 0, len(data) - 1, steps)

        output_filename = f"./output/merge_sort_step_{start_row + 1}_{end_row + 1}.txt"
        os.makedirs(os.path.dirname(output_filename), exist_ok=True)

        with open(output_filename, "w") as output_file:
            for step in steps:
                output_file.write(f"{step}\n")

        print(f"Merge Sort steps written to {output_filename}")

    except FileNotFoundError:
        print(f"Error: File '{filename}' not found.")
    except Exception as e:
        print(f"Error: {e}")


if __name__ == "__main__":
    main()
